xarray/core/nanops.py

```diff
@@ -0,0 +1,278 @@
+
+
+try:
+    import bottleneck as bn
+    _USE_BOTTLENECK = True
+except ImportError:
+    # use numpy methods instead
+    bn = np
+    _USE_BOTTLENECK = False
+
+try:
+    import dask.array as dask_array
+    from . import dask_array_compat
+except ImportError:
+    dask_array = None
+    dask_array_compat = None
+
+
+class bottleneck_switch(object):
+    """ xarray-version of pandas.core.nanops.bottleneck_switch """
+    def __call__(self, alt):
+        bn_name = alt.__name__
+
+        try:
+            bn_func = getattr(bn, bn_name)
+        except (AttributeError, NameError):  # pragma: no cover
+            bn_func = None
+
+        @functools.wraps(alt)
+        def f(values, axis=None, **kwds):
+            dtype = kwds.get('dtype', None)
+            min_count = kwds.get('min_count', None)
+
+            if (not isinstance(values, dask_array_type) and _USE_BOTTLENECK and
+                    not isinstance(axis, tuple) and
+                    values.dtype.kind in 'uifc' and
+                    values.dtype.isnative and
+                    (dtype is None or np.dtype(dtype) == values.dtype) and
+                    min_count is None):
+                # bottleneck does not take care dtype, min_count
+                kwds.pop('dtype', None)
+                kwds.pop('min_count', 1)
+                result = bn_func(values, axis=axis, **kwds)
+            else:
+                result = alt(values, axis=axis, **kwds)
+
+            return result
+
+        return f
+
+
+def _replace_nan(a, val):
+    """
+    If `a` is of inexact type, make a copy of `a`, replace NaNs with
+    the `val` value, and return the copy together with a boolean mask
+    marking the locations where NaNs were present. If `a` is not of
+    inexact type, do nothing and return `a` together with a mask of None.
+    Note that scalars will end up as array scalars, which is important
+    for using the result as the value of the out argument in some
+    operations.
+    Parameters
+    ----------
+    a : array-like
+        Input array.
+    val : float
+        NaN values are set to val before doing the operation.
+    Returns
+    -------
+    y : ndarray
+        If `a` is of inexact type, return a copy of `a` with the NaNs
+        replaced by the fill value, otherwise return `a`.
+    mask: {bool, None}
+        If `a` is of inexact type, return a boolean mask marking locations of
+        NaNs, otherwise return None.
+
+    This function is taken from
+    https://github.com/numpy/numpy/blob/v1.14.0/numpy/lib/nanfunctions.py
+    but slightly modified to take care of dask.array
+    """
+    if a.dtype == np.object_:
+        # object arrays do not support `isnan` (gh-9009), so make a guess
+        mask = a != a
+    elif issubclass(a.dtype.type, np.inexact):
+        mask = np.isnan(a)
+    else:
+        mask = None
+
+    if mask is not None:
+        if isinstance(a, dask_array_type):
+            return dask_array.where(mask, val, a), mask
```

Comment:
This is a little surprising to me. I would rather use `duck_array_ops.where()` here than use explicit type checks.

Likewise, instead of logic above for `mask`, use `duck_array_ops.isnull()`

Reply:
Done.